src/FsAutoComplete.Core/FileSystem.fs

```diff
@@ -385,10 +398,35 @@ module RoslynSourceText =
       member _.CopyTo(sourceIndex, destination, destinationIndex, count) =
         sourceText.CopyTo(sourceIndex, destination, destinationIndex, count)
 
+    interface ISourceTextNew with
+      member this.GetChecksum() = sourceText.GetChecksum()
+
 type ISourceTextFactory =
   abstract member Create: fileName: string<LocalPath> * text: string -> IFSACSourceText
   abstract member Create: fileName: string<LocalPath> * stream: Stream -> CancellableValueTask<IFSACSourceText>
 
+module SourceTextFactory =
+
+
+  let readFile (fileName: string<LocalPath>) (sourceTextFactory: ISourceTextFactory) =
+    cancellableValueTask {
+      let file = UMX.untag fileName
+
+      // use large object heap hits or threadpool hits? Which is worse? Choose your foot gun.
+
+      if FileInfo(file).Length >= File.bufferSize then
+        // Roslyn SourceText doesn't actually support async streaming reads but avoids the large object heap hit
+        // so we have to block a thread.
+        use s = File.openFileStreamForReadingAsync fileName
+        let! source = sourceTextFactory.Create(fileName, s)
+        return source
+      else
+        // otherwise it'll be under the LOH threshold and the current thread isn't blocked
+        let! text = fun ct -> File.ReadAllTextAsync(file, ct)
+        let source = sourceTextFactory.Create(fileName, text)
+        return source
+    }
```

Comment:
nice encapsulation of the decisions here, thanks for the comments

Reply:
Funny enough, I think this doesn't result in a good outcome. Blocking threads on smaller files is less of an issue since it'll be quick anyway.